mjpc/planners/model_derivatives.cc

```diff
@@ -48,40 +48,117 @@ void ModelDerivatives::Compute(const mjModel* m,
                                const double* h, int dim_state,
                                int dim_state_derivative, int dim_action,
                                int dim_sensor, int T, double tol, int mode,
-                               ThreadPool& pool) {
-  {
-    int count_before = pool.GetCount();
-    for (int t = 0; t < T; t++) {
-      pool.Schedule([&m, &data, &A = A, &B = B, &C = C, &D = D, &x, &u, &h,
-                     dim_state, dim_state_derivative, dim_action, dim_sensor,
-                     tol, mode, t, T]() {
-        mjData* d = data[ThreadPool::WorkerId()].get();
-        // set state
-        SetState(m, d, x + t * dim_state);
-        d->time = h[t];
-
-        // set action
-        mju_copy(d->ctrl, u + t * dim_action, dim_action);
+                               ThreadPool& pool, int skip) {
+  // reset indices
+  evaluate_.clear();
+  interpolate_.clear();
 
-        // Jacobians
-        if (t == T - 1) {
-          // Jacobians
-          mjd_transitionFD(m, d, tol, mode, nullptr, nullptr,
-                           DataAt(C, t * (dim_sensor * dim_state_derivative)),
-                           nullptr);
-        } else {
-          // derivatives
-          mjd_transitionFD(
-              m, d, tol, mode,
-              DataAt(A, t * (dim_state_derivative * dim_state_derivative)),
-              DataAt(B, t * (dim_state_derivative * dim_action)),
-              DataAt(C, t * (dim_sensor * dim_state_derivative)),
-              DataAt(D, t * (dim_sensor * dim_action)));
-        }
-      });
+  // evaluate indices
+  int s = skip + 1;
+  evaluate_.push_back(0);
+  for (int t = s; t < T - s; t += s) {
+    evaluate_.push_back(t);
+  }
+  evaluate_.push_back(T - 2);
+  evaluate_.push_back(T - 1);
+
+  // interpolate indices
+  for (int t = 0; t < T; t++) {
+    if (std::find(evaluate_.begin(), evaluate_.end(), t) == evaluate_.end()) {
```

Comment:
std::find makes the loop O(T^2), but it should be O(T) because evaluate_ is sorted, i.e.:

```
for (int t = 0, e = 0; t < T; t++) {
  if (e == evaluate_.size() || evaluate_[e] > t) {
    interpolate_.push_back(t);
  } else {
    e++;
  }
}
```

Reply:
Done.